cql3/restrictions/statement_restrictions.cc

```diff
@@ -392,8 +392,9 @@ std::vector<const column_definition*> statement_restrictions::get_column_defs_fo
                 }
             }
         }
-        if (_clustering_columns_restrictions->needs_filtering(*_schema)) {
-            column_id first_filtering_id = _schema->clustering_key_columns().begin()->id +
+        const bool pk_has_unrestricted_components = _partition_key_restrictions->has_unrestricted_components(*_schema) && !_partition_key_restrictions->is_on_token();
```

Comment:
I think this is correct, but can you please add a test for the is_on_token() case, that the optimization you allow here (not using filtering) is actually handled correctly?

Reply:
I will push the test in a followup to this pull request, crossing my fingers that it just adds a commit on top without destroying anything.

Reply:
I don't understand this. Isn't a restriction on the token equivalent to having unrestricted components?

ps. github "restricts" the view to about half my monitor width, forcing me to scroll horizonally. Thanks github.

Reply:
Is it? If I set a `token(pk) = 12345` restriction, it defines a partition. edit: Oh, but is_on_token() can be an inequality relation too... I see the point. I'll fix and send v2 tomorrow in the morning. It would make sense for equality relation only.

Reply:
No, multiple partitions can map to a single token.

Reply:
Right, it's a hash... My bad, I'll amend it tomorrow and purge these token "optimizations".

Reply:
I was also worried about this token thing, which is why I asked you to write a test for it, I forgot to check if your test used a range of tokens, not just a single token :-( At first I thought, like Avi, that you did it wrong, but then I wasn't sure... The reason why I thought it might actually work is that theoretically, Scylla's read code could accept both a partition token range *and* a clustering key range, and walk over all the partitions in that range and in each of the partitions, read the requested contiguous clustering-row range. I thought this was already supported, but maybe I was too optimistic. In any case, please test. If the test shows it doesn't work, no amount of theorizing can make it work :-)

Reply:
Ok. On the other hand, I was assigned to this issue with the remark that it has a strict deadline for Friday, so I think that tomorrow I'll start with the simpler fix which doesn't make any assumptions on how token restrictions work, and then, after the main bug is fixed, continue the research, because I don't want to get too sidetracked.

Reply:
Sounds good to me.